contracts/SharedSmartVault.sol

```diff
@@ -78,16 +92,123 @@ contract SharedSmartVault is AccessControlUpgradeable, UUPSUpgradeable, ERC4626U
       address(withdrawer_) != address(0),
       "SharedSmartVault: withdrawer_ cannot be zero address"
     );
+    require(investments_.length != 0, "SharedSmartVault: investments_ cannot be empty.");
     _setupRole(DEFAULT_ADMIN_ROLE, _msgSender());
 
     _collector = collector_;
     _withdrawer = withdrawer_;
-    // TODO: Infinite approval
+    for (uint256 i = 0; i < investments_.length; i++) {
+      this.addInvestment(investments_[i]);
+    }
+    // Infinite approval to the SmartVault
+    IERC20Metadata(asset()).approve(_smartVault, type(uint256).max);
   }
 
   // solhint-disable-next-line no-empty-blocks
   function _authorizeUpgrade(address newImpl) internal view override onlyRole(GUARDIAN_ROLE) {}
 
+  function _balance() internal view returns (uint256) {
+    return IERC20Metadata(asset()).balanceOf(address(this));
+  }
+
+  /**
+   * @dev See {IERC4626-deposit}.
+   */
+  function deposit(
+    uint256 assets,
+    address receiver
+  ) public virtual override onlyRole(LP_ROLE) returns (uint256) {
+    return super.deposit(assets, receiver);
+  }
+
+  function _deposit(
+    address caller,
+    address receiver,
+    uint256 assets,
+    uint256 shares
+  ) internal virtual override {
+    super._deposit(caller, receiver, assets, shares);
+    _collector.call(asset(), assets);
+    require(
+      _balance() == 0,
```

Comment:
Esto podría generarnos un ataque de DoS (Denial of Service). Si alguien nos transfiere 0.01 USDC, luego en un deposit de 1000, recibimos 1000 del LP (línea 130) y mandamos 1000 al SV (línea 131) y van a quedar igualmente 0.01 en el ShSV y va a revertir. 

Dos opciones: 

1. Cambiamos la validación para que se verifique que el balance entre las líneas 130 y 133 no cambia (en lugar de decir que sea 0). 
2. Cambiamos la línea 131 para que en lugar de assets envíe _balance()
3. Sacamos la validación.

Yo estoy más inclinado por la 1ra opción.